talisker/testing.py

```diff
@@ -0,0 +1,265 @@
+# Copyright (C) 2016- Canonical Ltd
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
+from builtins import *  # noqa
+import sys
+import subprocess
+import re
+import shlex
+import logging
+from contextlib import contextmanager
+import weakref
+
+import requests
+
+import talisker.logs
+
+
+class TestLogger(talisker.logs.StructuredLogger):
+
+    def __init__(self, name):
+        super().__init__(name)
+        self.records = []
+        self.lines = []
+
+
+class TestHandler(logging.Handler):
+    """Testing handler that records its logs in memory."""
+    def __init__(self, logger, level=logging.NOTSET):
+        assert isinstance(logger, TestLogger)
+        super().__init__(level)
+        self.logger = weakref.proxy(logger)
```

Comment:
I'm cheating by giving the handler a reference to it's owning Logger, which would be a reference cycle. The reason for that is so the handler can store the log messages on the logger.

There may be a cleaner way to do this. Instead of the reference, the handler could just store the log messages iself, and TestLogger could have a property that gets them from it's handlers.